### DISCUSSION TITLE
ESP32-S3 (Adafruit Feather board) bootloader randomly stuck in firmware download mode

### DISCUSSION THREAD

**Discussion by beriberikix** [UncleRus/esp-idf-lib#467] [Q&A] [answered]
I'm still feeling my way around FreeRTOS but I have a general understanding of kernel concepts. I currently have two sensors, `lc709203f` & `scd30`, correctly logging readings from a single task as seen [here](https://github.com/beriberikix/frog-esp-idf/blob/02642a6243094452a6d0c71243f369490b7ffe35/main/main.c#L68-L126). When I split them into two tasks ([here](https://github.com/beriberikix/frog-esp-idf/blob/split_task/main/main.c#L150-L151),) I only see "waiting for download" while running `idf.py monitor`, which is a sign things aren't working right.

Are tasks the right approach? Is there an example of a recommended way?

**Comment by UncleRus** [comment, marked as answer]
Your problem has nothing to do with FreeRTOS and programming. The bottom line is that the GPIO0 (BOOT0) pin must be pulled up to +3.3V in order for the bootloader to start executing the program and vice versa, if it is pulled to the ground, the bootloader waits for the firmware when the chip starts up, which happens in your case.

Your hardware has a bug, fix it by soldering a 10k resistor between GPIO0 and +3.3V.

**Reply by beriberikix** [reply]
Hmm...I'm using an Adafruit ESP32-S3 Feather and have been pouring over the [schematics](https://learn.adafruit.com/assets/110822). It looks like GPIO0 when not pressed is unconnected. And the docs [say](https://docs.espressif.com/projects/esptool/en/latest/esp32s3/advanced-topics/boot-mode-selection.html), "GPIO0 has an internal pullup resistor, so if it is left unconnected then it will pull high." So it should work, unless I'm misinterpreting. If that's the case, they don't expose GPIO0 for me to pull high :(

**Reply by UncleRus** [reply]
GPIO0 is connected to the "Boot" button.

**Reply by keenanjohnson** [reply]
Maybe I'm missing something, [but I don't understand how it can be a hardware bug when the docs clearly state that there is an internal pull up resistor? ](https://docs.espressif.com/projects/esptool/en/latest/esp32s3/advanced-topics/boot-mode-selection.html#gpio0). The docs even go on to say that connecting it to a button is quite common.

<img width="724" alt="image" src="https://user-images.githubusercontent.com/2559382/198882769-203cf7cd-c4a5-4534-bd00-99ae361c1992.png">

**Reply by trombik** [reply]
I don't understand why you ask us. we are not a manufacturer.

**Reply by keenanjohnson** [reply]
@UncleRus stated that he believed it to be a hardware bug, but I don't see how given the official docs. So I am trying to learn more about what I don't know.

**Reply by beriberikix** [reply]
I originally asked here because I believe it's a software bug, not hardware (still do.) But I'm pretty sure it's with the ESP-IDF tooling, not this library so I've reached out to Espressif. I'll update here if I learn anything.

**Reply by trombik** [reply]
@keenanjohnson the link you mentioned is espressif's doc, not the manufacturer of *your* board. if the chip is in download mode, it's not our problem. if you think otherwise, it is your responsibility to prove it. you haven't so far. 

I recently noticed the upload process has changed in recent esp-idf versions. I did not bother to investigate it because my primary development environment is a plain esp32, and it works as before.

**Reply by trombik** [reply]
for others, who have similar issues (the bootloader is in download mode):

if you are an engineer, try to isolate the problem. upload other examples in our library. try other chips, such as esp32, or esp32c3. try other examples in other libraries. measure the voltage on GPIO0.

if you are not, Arduino might be a better choice.

**Comment by UncleRus** [comment]
Same problem with same board here: #464
